helpers/stable_swap_math/mod.rs

```diff
@@ -0,0 +1,934 @@
+    let n = reserves.len() as u32;
+    let ann: U256 = casted_mul(
+        amp_coef,
+        n.checked_pow(n).ok_or(MathError::MulOverflow(9))?.into(),
+    );
+    let d: U256 = compute_d(reserves, amp_coef)?;
+
+    let mut c = d
+        .checked_mul(d)
+        .ok_or(MathError::MulOverflow(10))?
+        .checked_div(new_reserve_x.into())
+        .ok_or(MathError::DivByZero(3))?;
+    let mut reservers_sum: U256 = new_reserve_x.into();
+    // reserves_sum = ... + x_(i') + ...
+    // c1 = ... * d / x_(i') * ... * d
+    // where  i' in (0,n) AND i' != token_y_id
+    for (idx, &reserve) in reserves.iter().enumerate() {
+        if idx != token_x_id && idx != token_y_id {
+            reservers_sum = reservers_sum
+                .checked_add(reserve.into())
+                .ok_or(MathError::AddOverflow(5))?;
+            c = c
+                .checked_mul(d)
+                .ok_or(MathError::MulOverflow(11))?
+                .checked_div(reserve.into())
+                .ok_or(MathError::DivByZero(4))?;
+        }
+    }
+    // c = c_1 * d / (A * n^2n)
+    c = c
+        .checked_mul(d)
+        .ok_or(MathError::MulOverflow(12))?
+        .checked_div(
+            ann.checked_mul((n).checked_pow(n).ok_or(MathError::MulOverflow(13))?.into())
+                .ok_or(MathError::MulOverflow(14))?,
+        )
+        .ok_or(MathError::DivByZero(5))?;
+    // reserves_sum + d / ( A * n^n)
+    let b: U256 = d
+        .checked_div(ann)
+        .ok_or(MathError::DivByZero(6))?
+        .checked_add(reservers_sum)
+        .ok_or(MathError::AddOverflow(6))?; // d will be subtracted later
+
+    let mut y_prev = d;
+    let mut y = y_prev;
+    for _ in 0..MAX_ITERATIONS {
+        y = compute_y_next(y_prev, b, c, d)?;
+        if y > y_prev {
+            if y.checked_sub(y_prev).ok_or(MathError::SubUnderflow(4))? <= 1.into() {
+                return Ok(y.as_u128());
+            }
+        } else if y_prev.checked_sub(y).ok_or(MathError::SubUnderflow(5))? <= 1.into() {
+            return Ok(y.as_u128());
+        }
+        y_prev = y;
+    }
+    Ok(y.as_u128())
+}
+
+fn compute_y_next(y_prev: U256, b: U256, c: U256, d: U256) -> Result<U256, MathError> {
+    let numerator = y_prev
+        .checked_pow(2.into())
+        .ok_or(MathError::MulOverflow(15))?
+        .checked_add(c)
+        .ok_or(MathError::AddOverflow(7))?;
+    let denominator = y_prev
+        .checked_mul(2.into())
+        .ok_or(MathError::MulOverflow(16))?
+        .checked_add(b)
+        .ok_or(MathError::AddOverflow(8))?
+        .checked_sub(d)
+        .ok_or(MathError::SubUnderflow(6))?;
+    numerator
+        .checked_div(denominator)
+        .ok_or(MathError::DivByZero(7))
+}
+
+/// Compute SwapResult after an exchange given `amount_in` of the `token_in_id`.
+/// panics if token ids are out of bounds.
+/// NOTICE: it does not check if `token_in_id` != `token_out_id`.
+/// Returns (amount_out, fee_amount)
+pub fn swap_to(
```

Comment:
This doesn't need to be `pub`. It's used internally by `rated_swap_to`. It might look like a minor change but it actually lowers cognitive load on the reviewer (or at least for me).

Reply:
Done in [RM pub where unnecessary](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/5/commits/27df09c7891ee7a64f85b73e3da569b976652311).